Updates round 3

matthewhughes934 opened the pull request:
Final round of updates from my fork. After this I think all dependencies
should be close to their latest release, in my fork I added `dependabot`
after these changes to keep things up to date.

Context, other PRs were from 26cf27dcef72b2e2a99569b95462bd58bf1b7880,
78c30dc4ffee4c59fc8563ca1eab52624f57de60 and comment
https://github.com/PyCQA/isort/pull/2320#issuecomment-2579494172

- Update `mypy` to `v1`+

    This includes removing some now unused ignores, and fixing some new
    errors:

        isort/sorting.py:120: error: Unused "type: ignore" comment  [unused-ignore]
        isort/settings.py:869: error: Unused "type: ignore" comment  [unused-ignore]
        isort/literal.py:89: error: Unused "type: ignore" comment  [unused-ignore]
        isort/deprecated/finders.py:32: error: Incompatible types in assignment (expression has type "None", variable has type "Callable[[PathLike[Any], Optional[Any], bool, bool], Dict[str, Union[Requirement, UnparsedRequirement]]]")  [assignment]
        isort/api.py:196: error: No overload variant of "compile" matches argument types "str", "str", "str", "int", "int"  [call-overload]
        isort/api.py:196: note: Possible overload variants:
        isort/api.py:196: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: Literal[0], dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> CodeType
        isort/api.py:196: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, *, dont_inherit: bool = ..., optimize: int = ..., _feature_version: int = ...) -> CodeType
        isort/api.py:196: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: Literal[1024], dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> AST
        isort/api.py:196: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: int, dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> Any
        isort/api.py:223: error: No overload variant of "compile" matches argument types "str", "str", "str", "int", "int"  [call-overload]
        isort/api.py:223: note: Possible overload variants:
        isort/api.py:223: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: Literal[0], dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> CodeType
        isort/api.py:223: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, *, dont_inherit: bool = ..., optimize: int = ..., _feature_version: int = ...) -> CodeType
        isort/api.py:223: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: Literal[1024], dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> AST
        isort/api.py:223: note:     def compile(source: Union[str, Buffer, Module, Expression, Interactive], filename: Union[str, Buffer, _PathLike[Any]], mode: str, flags: int, dont_inherit: bool = ..., optimize: int = ..., *, _feature_version: int = ...) -> Any
        tests/integration/test_setting_combinations.py:64: error: Unused "type: ignore" comment  [unused-ignore]
        tests/unit/test_isort.py:5550: error: Signature of "seek" incompatible with supertype "TextIOWrapper"  [override]
        tests/unit/test_isort.py:5550: note:      Superclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, int, int = ..., /) -> int
        tests/unit/test_isort.py:5550: note:      Subclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, position: Any) -> Any
        tests/unit/test_isort.py:5550: error: Signature of "seek" incompatible with supertype "IOBase"  [override]
        tests/unit/test_isort.py:5550: note:      Superclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, int, int = ..., /) -> int
        tests/unit/test_isort.py:5550: note:      Subclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, position: Any) -> Any
        tests/unit/test_isort.py:5550: error: Signature of "seek" incompatible with supertype "IO"  [override]
        tests/unit/test_isort.py:5550: note:      Superclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, int, int = ..., /) -> int
        tests/unit/test_isort.py:5550: note:      Subclass:
        tests/unit/test_isort.py:5550: note:          def seek(self, position: Any) -> Any
        tests/integration/test_hypothesmith.py:69: error: Unused "type: ignore" comment  [unused-ignore]
        Found 11 errors in 8 files (checked 88 source files)

- Update `bugbear`

    And fix some issues from new `bugbear`:

        tests/unit/test_setuptools_command.py:15:5: B036 Don't except `BaseException` unless you plan to re-raise it.
        tests/unit/test_setuptools_command.py:30:5: B036 Don't except `BaseException` unless you plan to re-raise it.

matthewhughes934 commented:
@staticdev since I saw you merge my previous PR, I raised this just in case you wanted some more dependency updates since it was just a matter of cherry-picking from my fork

DanielNoord commented on isort/deprecated/finders.py:
```diff
@@ -29,7 +29,7 @@
 
 except ImportError:
-    parse_requirements = None
+    parse_requirements = None  # type: ignore[assignment]
```
This can be fixed by initialising `parse_requirements` on line 27 as `parse_requirements: types.ModelType | None` right? Or does `mypy` no longer support that pattern?

matthewhughes934 replied:
> This can be fixed by initialising `parse_requirements` on line 27 as `parse_requirements: types.ModelType | None` right? Or does `mypy` no longer support that pattern?

Sorry, what is `types.ModelType`?

I think the typing would have to be something like (from a glance at https://github.com/di/pip-api/blob/21eea42e37b29e18c004a922f8989108a006a203/pip_api/_parse_requirements.py#L484):

```python
parse_requirements: (
    Callable[
        [os.PathLike[str], Optional[Any], bool, bool],
        Dict[str, Union[pip_api.Requirement, pip_api.UnparsedRequirement]],
    ]
    | None
)
```

but that's a bit tricky, because at this point `pip_api` isn't imported, so `pip_api.UnparsedRequirement` is unknown

DanielNoord commented on pyproject.toml:
```diff
@@ -56,11 +56,11 @@ cruft = ">=2.12.0"
 hypothesmith = ">=0.1.3"
 hypothesis = ">=6.10.1"
-mypy = ">=0.902,<1.0.0"
+mypy = ">=0.902,<2.0.0"
```
Why limit this to below 2?

matthewhughes934 commented on pyproject.toml:
```diff
@@ -56,11 +56,11 @@ cruft = ">=2.12.0"
 hypothesmith = ">=0.1.3"
 hypothesis = ">=6.10.1"
-mypy = ">=0.902,<1.0.0"
+mypy = ">=0.902,<2.0.0"
```
> I would suggest to change to:

updated while rebasing on `main`: [01afb2e](https://github.com/PyCQA/isort/pull/2334/commits/01afb2e2b647f6af6b1de5bd09cabcc2cc14d8dc)

matthewhughes934 commented on pyproject.toml:
```diff
@@ -56,11 +56,11 @@ cruft = ">=2.12.0"
 example-shared-isort-profile = ">=0.1.0"
 flake8 = ">=3.8.4"
-flake8-bugbear = ">=22.12.6,<23.0.0"
+flake8-bugbear = ">=22.12.6,<25.0.0"
```
> Here I would do:

squashed in with my rebase: [a8dbecf](https://github.com/PyCQA/isort/pull/2334/commits/a8dbecf59ad73029a27be68ddf603221e382e94e)

matthewhughes934 commented:
actions look unhappy, but this is also true for `main`: checks are failing on https://github.com/PyCQA/isort/commit/1c8d963c54a6ed75d5f29e16835dcfba06bd12d3. I think https://github.com/PyCQA/isort/pull/2339 is addressing this

DanielNoord reviewed (approved):
LGTM!

@staticdev I think we can merge this already, it doesn't break `main` more than it already is

staticdev reviewed (changes requested):
@matthewhughes934 still we have to figure out what is the issue now with Python 3.13, since tests still are not passing.

matthewhughes934 commented:
> @matthewhughes934 still we have to figure out what is the issue now with Python 3.13, since tests still are not passing.

:+1: I've have a deeper look later today

matthewhughes934 commented:
> @matthewhughes934 still we have to figure out what is the issue now with Python 3.13, since tests still are not passing.

the failure was a built time, I've fixed it up and rebased, see commit https://github.com/PyCQA/isort/pull/2334/commits/faf22ce9a0018dc25d14d61ac3a5ea1692032ea8 for details (the last bit of the body)

staticdev commented:
Sorry @matthewhughes934 due to a merge of new Poetry this lock got obsolete. Could you please rebase and lock again?

matthewhughes934 commented:
> Sorry @matthewhughes934 due to a merge of new Poetry this lock got obsolete. Could you please rebase and lock again?

:+1: looks like that change brought in all the updates from my first commit, so that's gone away with the rebase

staticdev commented:
LGTM, thanks @matthewhughes934 @DanielNoord
